### DISCUSSION TITLE
"supabase_functions.http_request" does not exist when called in function

### DISCUSSION THREAD

**Discussion by stefan-girlich** [supabase/supabase#22668] [Questions] [answered]
When a user registers I want to trigger sending the corresponding email via a REST API. This is not supported by Supabase, so I want to achieve this via a webhook calling an edge function. Webhooks with edge functions are not supported in a local setup and require hard-coding URLs and credentials, so I want to use a custom function which accesses credentials from `vault`.

I would appreciate any hint, thanks!

## Function and hook
```sql
CREATE OR REPLACE FUNCTION public.trigger_edge_function()
RETURNS TRIGGER AS $$
DECLARE
    base_url TEXT;
    service_key TEXT;
    request_url TEXT;
    auth_header jsonb;
BEGIN
    SELECT decrypted_secret
    INTO base_url
    FROM vault.decrypted_secrets
    WHERE name = 'supabase_functions_base_url';

    SELECT decrypted_secret
    INTO service_key
    FROM vault.decrypted_secrets
    WHERE name = 'service_role_key';

    request_url := base_url || '/dummy';
    auth_header := '{"Content-type":"application/json","Authorization":"Bearer ' || service_key || '"}';

    select supabase_functions.http_request(
        request_url,
        'POST',
        auth_header,
        '{}',
        '1000'
    );

    RETURN NEW;
END;
$$ LANGUAGE plpgsql
SECURITY DEFINER
SET search_path = supabase_functions, extensions, public, pg_temp;

CREATE OR REPLACE TRIGGER dummy_hook
AFTER INSERT ON public.my_users_table
FOR EACH ROW
EXECUTE FUNCTION public.trigger_edge_function();
```

## Error
```
ERROR: 42883: function supabase_functions.http_request(text, unknown, jsonb, unknown, unknown) does not exist
HINT: No function matches the given name and argument types. You might need to add explicit type casts.
```

## Further observations
The following webhook/trigger created via UI works as expected:
```sql
CREATE TRIGGER foo AFTER INSERT ON public.my_users_table FOR EACH ROW EXECUTE FUNCTION supabase_functions.http_request('http://host.docker.internal:54321/functions/v1/dummy', 'POST', '{"Content-type":"application/json"}', '{}', '1000')
```

The following standalone statement fails with the original error (except all fields are "unknown" in the function signature):
```sql
select supabase_functions.http_request('http://host.docker.internal:54321/functions/v1/dummy', 'POST', '{"Content-type":"application/json"}', '{}', '1000')
```

**Comment by GaryAustin1** [comment, marked as answer]
Why don't you just use the http or pg_net extension instead of trying to get the webhook code involved.

**Reply by stefan-girlich** [reply]
I was hoping to use an existing function to avoid re-writing. Thanks for the hint; I ended up copy-pasting the [original function definition](https://github.com/supabase-community/supabase-graphql-example/blob/main/data/db/schema.sql#L1261) end it worked as expected.

**Reply by DreamwareDevelopment** [reply]
Same, Supabase can be a pain to work with sometimes.

**Reply by vineethvijayan314** [reply]
@stefan-girlich or @DreamwareDevelopment  can you explain how did you fix it? Did you remove the part of Returns Trigger from the original function?

**Comment by mirzafuzail123** [comment]
supabase_functions.http_request is a trigger function so you cannot call it inside another function. It has to be executed directly on your webhook trigger something like this:


CREATE TRIGGER dummy_webhook
AFTER UPDATE
ON public.dummy_table
FOR EACH ROW WHEN (OLD.name IS DISTINCT FROM NEW.name)
EXECUTE FUNCTION supabase_functions.http_request(
            '', -- Webhook URL
            'POST',
            '{"Content-Type":"application/json"}',
            '{}',
            '1000'-- Timeout in milliseconds
        );

**Comment by Rubenburdin** [comment]
Supabase_functions.http_request is defined as a trigger function (returns event_trigger or similar), which means Postgres only allows it in the EXECUTE FUNCTION clause of a trigger definition - you can't call it with SELECT from inside another function.
The cleanest fix is to use net.http_post from the pg_net extension instead, which is designed to be called from anywhere Here's roughly what the refactored function would look like:

CREATE OR REPLACE FUNCTION public.trigger_edge_function()
RETURNS TRIGGER AS $$
DECLARE
    base_url TEXT;
    service_key TEXT;
BEGIN
    SELECT decrypted_secret INTO base_url
    FROM vault.decrypted_secrets WHERE name = 'supabase_functions_base_url';

    SELECT decrypted_secret INTO service_key
    FROM vault.decrypted_secrets WHERE name = 'service_role_key';

    PERFORM net.http_post(
        url := base_url || '/dummy',
        headers := jsonb_build_object(
            'Content-Type', 'application/json',
            'Authorization', 'Bearer ' || service_key
        ),
        body := jsonb_build_object('user_id', NEW.id),
        timeout_milliseconds := 5000
    );

    RETURN NEW;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;

This approach also has the benefit of being more flexible - you can dynamically build the URL, headers, and body based on the trigger context, which you can't really do with the webhook UI.
One thing to watch: pg_net is asynchronous, so youur trigger won't wait for the Edge Function response. If you need synchronous behavior (rare, but sometimes necessary), you'd use the http extension instead with http_post(). But for most cases the async pattern is preferable since it won't block your insert.
